src/goto-instrument/contracts/assigns.cpp

```diff
@@ -197,47 +183,16 @@ goto_programt assigns_clauset::dead_stmts()
 
 goto_programt assigns_clauset::havoc_code()
 {
-  goto_programt havoc_statements;
-  source_locationt location = assigns.source_location();
-
-  for(assigns_clause_targett *target : targets)
-  {
-    // (1) If the assigned target is not a dereference,
-    // only include the havoc_statement
-
-    // (2) If the assigned target is a dereference, do the following:
-
-    // if(!__CPROVER_w_ok(target, 0)) goto z;
-    //      havoc_statements
-    // z: skip
+  modifiest modifies;
+  std::for_each(
+    targets.begin(),
+    targets.end(),
+    [&modifies](const assigns_clause_targett *t) {
+      modifies.insert(to_address_of_expr(t->get_direct_pointer()).object());
+    });
```

Comment:
Very clean 👏 nice!

Reply:
😄 Thanks!

While writing this transformation, I noticed that `targets` is a `vector`, unlike `modifies`, which is a `set`. If we write the same expression twice in the `__CPROVER_assigns` clause, would we end up havocing / enforcing it twice? Or, does the parser / typechecker somehow guard against this?

Reply:
> If we write the same expression twice in the `__CPROVER_assigns` clause, would we end up havocing / enforcing it twice? Or, does the parser / typechecker somehow guard against this?

We don't have any regression tests that cover this behavior. Could you write a regression test that cover this scenario for both loop contracts and function contracts?

> While writing this transformation, I noticed that `targets` is a `vector`, unlike `modifies`, which is a `set`.

Could you turn `targets` into a `set` too? Regardless of our type-checking procedure, `set` would be a more appropriate data structure here.

These modifications might be part of a separate PR. I'm happy to implement it too if that's the case.

Reply:
Currently this is not a concern for loops because we don't allow manual `assigns` clause annotation yet.

I would separate this refactor (`vector` -> `set`) into another PR though. I'll open an internal ticket to track it.